clippy_lints/src/no_effect.rs

```diff
@@ -87,6 +87,15 @@ impl<'tcx> LateLintPass<'tcx> for NoEffect {
 
 fn check_no_effect(cx: &LateContext<'_>, stmt: &Stmt<'_>) -> bool {
     if let StmtKind::Semi(expr) = stmt.kind {
+        // move `expr.span.from_expansion()` ahead
+        if expr.span.from_expansion() {
+            return false;
+        }
+        let expr = peel_blocks(expr);
+        // assume nontrivial oprand of `Binary` Expr can skip `check_unnecessary_operation`
+        if is_operator_overriden(cx, expr) {
+            return true;
+        }
```

Comment:
NIT: I find it a bit weird that this line also influences `unnecessary_operation` but I have no idea how to rewrite it better. I think this adjustment will make it a bit clearer and then it should be fine:

```suggestion
        if is_operator_overridden(cx, expr) {
            // Return `true`, to prevent `check_unnecessary_operation` from
            // linting on this statement as well.
            return true;
        }
```

Could you also add tests to the file for `unnecessary_operation` to ensure that it doesn't lint on this expression?